specs/fjord/exec-engine.md

```diff
@@ -0,0 +1,114 @@
+
+#### Fjord L1-Cost fee changes (FastLZ estimator)
+
+Fjord updates the L1 cost calculation function to use a FastLZ-based compression estimator.
+The L1 cost is computed as:
+
+```pseudocode
+l1FeeScaled = l1BaseFeeScalar*l1BaseFee*16 + l1BlobFeeScalar*l1BlobBaseFee
+l1CostSigned = (intercept + fastlzCoef*fastlzSize + txSizeCoef*txSize) * l1FeeScaled / 1e12
+l1Cost = uint256(max(0, l1CostSigned))
+```
+
+| Input arg         | Type      | Description                                                       |
+|-------------------|-----------|-------------------------------------------------------------------|
+| `l1BaseFee`       | `uint256` | L1 base fee of the latest L1 origin registered in the L2 chain    |
+| `l1BlobBaseFee`   | `uint256` | Blob gas price of the latest L1 origin registered in the L2 chain |
+| `fastlzSize`      | `uint256` | Length of the FastLZ-compressed RLP-encoded signed tx             |
+| `txSize`          | `uint256` | Length of uncompressed RLP-encoded signed tx                      |
+| `l1BaseFeeScalar` | `uint32`  | L1 base fee scalar, scaled by `1e6`                               |
+| `l1BlobFeeScalar` | `uint32`  | L1 blob fee scalar, scaled by `1e6`                               |
+| `intercept`       | `int32`   | Intercept constant, scaled by `1e6` (can be negative)             |
+| `fastlzCoef`      | `int32`   | FastLZ coefficient, scaled by `1e6` (can be negative)             |
+| `txSizeCoef`      | `int32`   | Transaction size coefficient, scaled by `1e6` (can be negative)   |
+
+Where:
+
+- the `intercept` is hard-coded as `-27_321_890`
+
+- the `fastlzCoef` is hard-coded as `1_031_462`
+
+- the `txCoef` is hard-coded as `-88_664`
+
+- the `l1CostSigned` calculation is an unlimited precision signed integer computation, with the result in Wei and
+  having `int256` range.
+
+- the final `l1Cost` computation is an unlimited precision unsigned integer computation, with the result in Wei and
+  having `uint256` range.
+
+##### L1-Cost linear regression details
+
+The `intercept`, `fastlzCoef`, and `txSizeCoef` constants are calculated by linear regression using a dataset
+of previous L2 transactions. The dataset is generated by iterating over all transactions in a given time range, and
+performing the following actions. For each transaction:
+
+1. Calculate the RLP-encoded transaction payload. Record the length of this payload as `txSize`.
+2. Compress the payload using FastLZ. Record the length of the compressed payload as `fastlzSize`.
+3. Compress the payload using a "best estimate" compression, defined as:
+
+- Create two zlib streams, both using the maximum compression level.
+- Bootstrap one of the streams with at least 64kb of transaction data (after compression).
+- Write each transaction to both streams (flushing after each write), and record the size increase
+  of the larger buffer as `bestEstimateSize`.
+- Each time the larger buffer grows larger than 128kb, empty the stream and swap the buffers.
```

Comment:
FWIW Now that we batch 6 blobs at a time we found that you get about 2% more compression, which is probably not enough difference to affect the results much, but wondering if we should confirm that by increasing this to 128kb*6.